### PR TITLE
Vanadis - Refactoring codebase to modularize instruction and LSQ definitions

### PR DESCRIPTION
**Pull request by Anunalla**
VanadisInstruction, VanadisBasicLSQ and decoder classes were refactored to improve code reuse and readability. The current class design also makes it easier to introduce SIMT pipeline capabilities in the future (a separate PR is planned for the proposed SIMT feature).

### CODE REVIEW

File: src/sst/elements/vanadis/inst/regfile.h [hunk 1]
```diff
@@ -1,8 +1,8 @@
-// Copyright 2009-2024 NTESS. Under the terms
+// Copyright 2009-2023 NTESS. Under the terms
 // of Contract DE-NA0003525 with NTESS, the U.S.
 // Government retains certain rights in this software.
 //
-// Copyright (c) 2009-2024, NTESS
+// Copyright (c) 2009-2023, NTESS
```
Comment (hughes-c) [inline src/sst/elements/vanadis/inst/regfile.h:5]: Can you change these back to 2024?

File: src/sst/elements/vanadis/inst/regfile.h [hunk 2]
```diff
@@ -82,50 +90,63 @@ class VanadisRegisterFile
         assert(reg < count_fp_regs);
         assert((offset + len) <= fp_reg_width);
 
-        uint8_t* reg_ptr = (uint8_t*) &fp_reg_storage[reg * fp_reg_width];
+        int index = get_reg_index(reg,1);
+        uint8_t* reg_ptr = (uint8_t*) &fp_reg_storage[index];
+        // output->verbose(CALL_INFO, 16, 0, "CopyFromFPReg: reg=%d fp_reg_width=%d target_tid=%d WARP_SIZE=%d count_fp_regs=%d index = %d\n",
```
Comment (hughes-c) [inline src/sst/elements/vanadis/inst/regfile.h:95]: Please clean up the debug statements

File: src/sst/elements/vanadis/inst/regfile.h [hunk 3]
```diff
@@ -82,50 +90,63 @@ class VanadisRegisterFile
...
 
     void copyFromIntRegister(uint16_t reg, uint32_t offset, uint8_t* values, uint32_t len) {
         assert(reg < count_int_regs);
         assert((offset + len) <= int_reg_width);
-
-        uint8_t* reg_ptr = (uint8_t*) &int_reg_storage[reg * int_reg_width];
+        
+        int index = get_reg_index(reg, 0);
+        uint8_t* reg_ptr = (uint8_t*) &int_reg_storage[index];
+        // output->verbose(CALL_INFO, 16, 0, "CopyFromIntReg: reg=%d int_reg_width=%d target_tid=%d WARP_SIZE=%d count_int_regs=%d index= %d len=%d\n",
```
Comment (hughes-c) [inline src/sst/elements/vanadis/inst/regfile.h:112]: More debugging; please check all of the files for code that has been commented-out for deubgging

File: src/sst/elements/vanadis/os/include/futex.h [hunk 4]
```diff
@@ -36,7 +36,7 @@ namespace OS {
 class Futex {
 public:
     void addWait( uint64_t addr, VanadisSyscall* syscall ) {
-        OSFutexDbg("Futex::%s() addr=%#" PRIx64 " syscall=%p\n", addr,syscall );
+        // OSFutexDbg("Futex::%s() addr=%#" PRIx64 " syscall=%p\n", addr,syscall );
```
Comment (hughes-c) [inline src/sst/elements/vanadis/os/include/futex.h:39]: Why were these replaced with printf's?

File: src/sst/elements/vanadis/os/syscall/exit.cc [hunk 5]
```diff
@@ -44,8 +44,12 @@ VanadisExitSyscall::VanadisExitSyscall( VanadisNodeOSComponent* os, SST::Link* c
         m_buffer.resize(sizeof(uint32_t));
         *((uint32_t*)m_buffer.data()) = 0;
         writeMemory( tid_addr, m_buffer );
+        m_output->verbose(CALL_INFO, 16, VANADIS_OS_DBG_SYSCALL, "[syscall-exit] core %d thread %d process %d writeMemory() called \n",event->getCoreID(), event->getThreadID(), process->getpid() );
     } else {
+        m_output->verbose(CALL_INFO, 16, VANADIS_OS_DBG_SYSCALL, "[syscall-exit] core %d thread %d process %d memReqIsDone(true) called \n",event->getCoreID(), event->getThreadID(), process->getpid() );
+        printf("[syscall-exit] core %d thread %d process %d memReqIsDone(true) called \n",event->getCoreID(), event->getThreadID(), process->getpid());
```
Comment (hughes-c) [inline src/sst/elements/vanadis/os/syscall/exit.cc:50]: Do you need this printf? The inclusion of printf's will likely cause the autotester to fail if the outputs do not match.

File: src/sst/elements/vanadis/os/syscall/futex.cc [hunk 6]
```diff
@@ -123,13 +123,14 @@ VanadisFutexSyscall::VanadisFutexSyscall( VanadisNodeOSComponent* os, SST::Link*
 
 void VanadisFutexSyscall::wakeWaiter(VanadisSyscallFutexEvent* event) const
 {
+    printf("Calling findFutex from wakeWaiter\n");
```
Comment (hughes-c) [inline src/sst/elements/vanadis/os/syscall/futex.cc:126]: Please grep for all of your printf additions and ensure that they aren't left over from debugging.

File: src/sst/elements/vanadis/vanadis.cc [hunk 7]
```diff
@@ -436,16 +435,19 @@ VANADIS_COMPONENT::setHalt(uint32_t thr, int64_t halt_code)
     output->verbose(
         CALL_INFO, 2, 0, "-> Receive halt request on thread %" PRIu32 " / code: %" PRId64 "\n", thr, halt_code);
 
-    if ( thr >= hw_threads ) {
-        // Incorrect thread, ignore? error?
-    }
-    else {
+    // if ( thr >= hw_threads ) {
```
Comment (hughes-c) [inline src/sst/elements/vanadis/vanadis.cc:438]: Is this fixed? Can we remove this block?

File: src/sst/elements/vanadis/vanadis.h [hunk 8]
```diff
@@ -245,7 +252,8 @@ class VanadisComponent : public SST::Component
     int  allocateFunctionalUnit(VanadisInstruction* ins);
     bool mapInstructiontoFunctionalUnit(VanadisInstruction* ins, std::vector<VanadisFunctionalUnit*>& functional_units);
     void printRob(int rob_num, VanadisCircularQueue<VanadisInstruction*>* rob);
-
+    // bool judgeIns(VanadisInstruction* ins);
```
Comment (hughes-c) [inline src/sst/elements/vanadis/vanadis.h:255]: ?

File: src/sst/elements/vanadis/vanadisDbgFlags.h [hunk 9]
```diff
@@ -16,11 +16,18 @@
 #ifndef _H_VANADIS_DBG_FLAGS
 #define _H_VANADIS_DBG_FLAGS
 
-#define VANADIS_DBG_INS_LDR_FLG  (1<<0)
-#define VANADIS_DBG_DECODER_FLG  (1<<1)
-#define VANADIS_DBG_ISSUE_FLG  (1<<2)
-#define VANADIS_DBG_CYCLE_FLG  (1<<3)
-#define VANADIS_DBG_RETIRE_FLG  (1<<4)
-#define VANADIS_DBG_CHECKPOINT  (1<<5)
+// #define VANADIS_DBG_INS_LDR_FLG  (1<<0)
```
Comment (hughes-c) [inline src/sst/elements/vanadis/vanadisDbgFlags.h:19]: Why were these changed?

**Review by hughes-c — CHANGES_REQUESTED** [review]

File: src/sst/elements/vanadis/vanadis.cc (hunk 7 above)
Comment (Anunalla) [reply src/sst/elements/vanadis/vanadis.cc:438]: This was removed when we were creating new threads to act as warps. And then we changed to using existing threads as warps after which this change was leftover. I will revert to the original.

**Review by hughes-c — APPROVED** [review]

**Comment by hughes-c** [comment]
I will need to dig into your changes more because they changed the output of the test suite. I can't list everything but:
`node0.cpu0:decoder0:branch_unit.branch_cache_hit.1 : Accumulator : Sum.u64 = 170720; SumSQ.u64 = 170720; Count.u64 = 170720; Min.u64 = 1; Max.u64 = 1;` -->
`node0.cpu0:decoder0:branch_unit.branch_cache_hit.1 : Accumulator : Sum.u64 = 177134; SumSQ.u64 = 177134; Count.u64 = 177134; Min.u64 = 1; Max.u64 = 1; `

**Comment by hughes-c** [comment]
@Anunalla It looks like the differences are small enough to overwrite the reference files with the new values. But, there's also an issue with the osx (llvm) compilation. 
```
In file included from ./inst/vinstall.h:41:
./inst/vxor.h:44:18: warning: 'printToBuffer' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    virtual void printToBuffer(char* buffer, size_t buffer_size)
                 ^
./inst/vinst.h:319:22: note: overridden virtual function is here
        virtual void printToBuffer(char* buffer, size_t buffer_size) { snprintf(buffer, buffer_size, "%s", getInstCode()); }
                     ^
In file included from decoder/vmipsdecoder.cc:19:
In file included from ./decoder/vmipsdecoder.h:22:
In file included from ./inst/vinstall.h:30:
./inst/vmuli.h:53:69: error: non-virtual member function marked 'override' hides virtual member function
    void execute(SST::Output* output, VanadisRegisterFile* regFile) override
```

**Review by hughes-c — APPROVED** [review]
